src/endpoint/s3/s3_bucket_policy_utils.js

```diff
@@ -189,7 +201,23 @@ function _is_principal_fit(account, statement) {
             break;
         }
     }
-    return statement.Principal ? principal_fit : !principal_fit;
+
+    if (statement.NotPrincipal) {
+        // If the account is categorized as "Not Principal," we need to ALLOW operations
+        // for that account under the associated principal.
+        //
+        // To maintain backward compatibility, we also support name-based comparison for principal matching.
+        //
+        // - If `account_identifier_type` is "id" but the `account` parameter contains a name,
+        //   we return `false` to proceed/bypass to the next step, where we check for a name match.
+        //
+        // - In the next step, if `account_identifier_type` is "name" and the `account` also contains a name,
+        //   we return the actual value.
```

Comment:
What if the allow is by ids and the deny is by names?
You can add it a test case.
What will happen when we support ARN, it will be more complex.

Reply:
The name is another option, we decided to not remove this to be backward compatible.
Anyway, it is a valid option, and in the containerized it is the only option.